pep-0677.rst

```diff
@@ -635,24 +635,92 @@ callable types and ``=>`` for lambdas.
 Runtime Behavior
 ----------------
 
-Our tentative plan is that:
+The new AST nodes need to evaluate to runtime types, and we have two goals for the
+behavior of these runtime types:
 
-- The ``__repr__`` will show an arrow syntax literal.
-- We will provide a new API where the runtime data structure can be
-  accessed in the same manner as the AST data structure.
-- We will ensure that we provide an API that is backward-compatible
-  with ``typing.Callable`` and ``typing.Concatenate``, specifically
-  the behavior of ``__args__`` and ``__parameters__``.
+- They should expose a structured API as similar as possible to the AST where
+  the parameter and return types have descriptive names.
```

Comment:
The flattest possible structure would be to make it all one big structure with these fields
- `is_async`: bool
- `is_any_params`: bool
-  `positional_only_args`: tuple[typing.Type] | None
- `param_spec`: typing.ParamSpec | None
plus the backward compatible `__args__` and `__parameters__`.

I could make the fields use dunders, would rather decide that later since it's easy to change.

Whether this is a good approach or not is unclear. It is pythonic in some sense, since historically python had no static typing and so it wasn't possible to do much better than a bunch of fields, some combinations of which are illegal.

But going forward we do have static typing and pattern matching, which means that using an API where the structure of the data tells us which combinations of field values are legal we can do better. And if we eventually have extended syntax, the same concern would apply which does affect the API we want today.

Specifically these are the restrictions I mean:
- Today, having a separate representation for AnyArguments is nice, because then a pattern-matching client (or pre 3.10 a client using refinement) will know that we can't have positional only arguments and any arguments at the same time
- If we propose extended syntax in the future, having a separate representation for Concatenate versus ArgumentsList will allow a client using pattern matching / refinement to see that we can either have a `param_spec` or we can have values for `vararg` and `kwarg` but we cannot have both

So if we do want to pick an API where clients can use refinement or pattern matching, it pushes us toward something very similar to the AST. You're definitely right that we can make `is_async` just a flag with no downsides.

We don't have to have this though, we could expose a flat representation which would be easier to implement; I don't actually maintain any libraries that use runtime types so I don't much care for my own sake. If I did, I'd probably prefer the approach that allows pattern matching but I'm probably biased by spending about 1/4 of my time in ocaml.

Reply:
I think what it comes down to is that we can make many complete representations, but that which ones are confusing will depend on whether typical users care about either a pattern matching or type refinement friendly API.

If yes, then the more AST-like approach where we switch on types is the less confusing option. If no, then the flatter option with many fields is probably less confusing.

My guess is that since users of this API will be people working on typing-oriented libraries, they'd prefer the more structured and AST-like approach but I might be wrong

Reply:
I wrote up an untested version of how my pyanalyze library might parse either approach: https://gist.github.com/JelleZijlstra/91d31f04e6eea5bb24b96f06c06fedb2. I don't think there's a big difference between the two, and in either case it's a small amount of work to parse the syntax. However, the flat approach requires fewer imports and has fewer names to remember. On the other hand, it makes a couple of illegal states at least theoretically possible (like having both `is_any_params` and `param_spec` set).

I feel like a more important consideration will be how easy it's going to be to extend the syntax to more complicated callables, if we decide to move forward with that. For example, if we add default argument support, the nested approach would require changing both CallableConcatenation and CallableArgumentsList, while the flat approach would require changing only one class.

Reply:
Yeah, it's definitely fewer changes to use the flat approach - it's uniformly easier to implement and maintain.

But the number of illegal states grows, so the benefits of the nested approach get bigger in some sense as the signatures become more complicated.

I'd actually defer to your opinion here since I'm unlikely to ever use the runtime types - if from pyanalyze's point of view you don't see any difference, I'm happy to switch it to the flat approach. It will certainly make my life easier implementing this!

Reply:
I think it doesn't matter too much to me; I don't really care if it's 20 or 30 lines I have to write once. The only thing that's really important is that all the information about the type remains available. For example, the original implementation of ParamSpec had `P.args` as just `object()`, which would have made it impossible for pyanalyze to understand `P.args`. None of what you're suggesting here would have such problems, so that's enough to keep me happy.

I mildly prefer the flat approach though; it gives me fewer type names I have to remember, and the Zen of Python does say "flat is better than nested" :).

Reply:
Sounds good, I'll update to a flat interface.

I'll take a look at `inspect` and see if I can use compatible field names and types